src/AccessibilityInsights.VersionSwitcher/InstallationEngine.cs

```diff
@@ -0,0 +1,281 @@
+        /// </summary>
+        /// <param name="localFile">The full path to the local file</param>
+        /// <returns>An initialized TrustVerifier object</returns>
+        private static TrustVerifier ValidateLocalFile(string localFile)
+        {
+            TrustVerifier verifier = new TrustVerifier(localFile);
+            if (!verifier.IsVerified)
+            {
+                // TODO : Better error messaging
+                throw new ArgumentException("Untrusted file!", nameof(localFile));
+            }
+
+            EventLogger.WriteInformationalMessage("Successfully validated local file: {0}", localFile);
+
+            return verifier;
+        }
+
+        /// <summary>
+        /// Wrap the install inside a transaction
+        /// </summary>
+        /// <param name="options">The parameters to drive the installation</param>
+        /// <param name="transaction">The transaction to wrap the operation</param>
+        private void InstallWithinTransaction(InstallationOptions options, Transaction transaction)
+        {
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            bool success = false;
+            try
+            {
+                RemoveOldVersion();
+                InstallNewVersion(options.LocalInstallerFile);
+                success = true;
+            }
+            finally
+            {
+                if (success)
+                {
+                    transaction.Commit();
+                }
+                else
+                {
+                    transaction.Rollback();
+                }
+
+                stopwatch.Stop();
+
+                const string messageTemplate = "VersionSwitcher {0} in {1} ms";
+
+                if (success)
+                {
+                    EventLogger.WriteInformationalMessage(messageTemplate, "succeeded", stopwatch.ElapsedMilliseconds);
+                }
+                else
+                {
+                    EventLogger.WriteErrorMessage(messageTemplate, "failed", stopwatch.ElapsedMilliseconds);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Update the config file to reflect the new channel setting
+        /// </summary>
+        /// <param name="newChannel">The new channel to use</param>
+        private static void UpdateConfigWithNewChannel(string newChannel)
+        {
+            if (newChannel != null)
+            {
+                var configFile = Path.Combine(DirectoryManagement.sConfigurationFolderPath, Constants.AppConfigFileName);
+
+                SettingsDictionary settings = FileHelpers.LoadDataFromJSON<SettingsDictionary>(configFile);
+                settings[Constants.ReleaseChannelKey] = newChannel;
+                FileHelpers.SerializeDataToJSON(settings, configFile);
+            }
+        }
+
+        /// <summary>
+        /// Locate the MSI-installed app and launch it without elevated privileges
+        /// </summary>
+        private void LaunchPostInstallApp()
```

Comment:
the existing MSI usually launches itself upon installation already; what is the thinking behind this method?

Reply:
The launch on install doesn't trigger when we install the MSI through the API. I'm not really sure why that's true, but it ultimately works to our advantage. Since we are elevated when we install the MSI, the auto-launched app would also be elevated (this is easy to confirm by starting the MSI directly from an elevated command prompt, then saving a test file to C:\). We call LaunchPostInstallApp to launch AIWin after we make the change, and we launch it indirectly through explorer.exe to run it in a non-elevated state.